config/300-gateway.yaml

```diff
@@ -48,8 +48,13 @@ spec:
             - --base-id 1
             - -c /tmp/config/envoy-bootstrap.yaml
             - --log-level info
+            - --drain-time-s $(DRAIN_TIME_SECONDS)
```

Comment:
Envoy in Istio uses the same idea https://github.com/istio/istio/blob/master/pkg/envoy/proxy.go#L128 the question is what is the best option compared to the current one.


In envoy docs I read that two of the cases where draining happens is when:

The server begins the graceful drain sequence via the [drain_listeners?graceful](https://www.envoyproxy.io/docs/envoy/latest/operations/admin#operations-admin-interface-drain) admin endpoint.

The server has been manually health check failed via the [healthcheck/fail](https://www.envoyproxy.io/docs/envoy/latest/operations/admin#operations-admin-interface-healthcheck-fail) admin endpoint. See the [health check filter](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/health_checking#arch-overview-health-checking-filter) architecture overview for more information.

So I am wondering if current approach with posting to `http://localhost/healthcheck/fail` would work if we make the sleep time configurable. We probably need to add a test for this with some reproducer if possible.

Reply:
Thanks for the review. I've tested both; in the `preStop` hook, calling `/drain_listeners?graceful` is similar to calling `/healthcheck/fail`, in the sense both are starting the draining process as you have seen in [Envoy docs](https://www.envoyproxy.io/docs/envoy/v1.26.0/intro/arch_overview/operations/draining#arch-overview-draining). The only difference is that the `/ready` endpoint will now return `503` in the `/healthcheck/fail` case, which is not _that_ important because the pod is already in `Terminating` status when the `preStop` hook is called. 

On **my side**, I feel like `/drain_listeners?graceful` is somewhat clearer than `/healthcheck/fail`, because the former clearly states the goal is to drain, while the latter just makes the healthcheck fail, and people have to read Envoy's doc to know that failing the healthcheck causes starting the drain process.

Anyway, theorically, we can just do the following in the `preStopHook`:

- call `/healthcheck/fail` (or `/drain_listeners?graceful`, similar as I've seen)
- wait (sleep) during `<drain_max_time>` **OR** wait until there is no more in-flight requests

Waiting until all in-flight requests are completed is not simple. We could hack around [like I did](https://github.com/knative-extensions/net-kourier/pull/1203/commits/3e3e933c1100a4c82c751b045a0e5b8678a4a972), but I'm not sure it's the most robust/simple-to-understand solution. So, it might just be easier to wait using `sleep`. The downside of `sleep`ing is that we might wait too much (e.g. all in-flight requests can be completed after 10 seconds, but if we have configured a `<drain_max_time>` of 5 minutes, we will wait 4m50s for nothing). Both solutions have pros and cons.

Another thing we have to be careful is Kubernetes' `terminationGracePeriodSeconds`, that will automatically kill the pod after that period (and it applies also for the `preStopHook`). So, if we have `drain_max_time = 60s`, then `terminationGracePeriodSeconds` must be greater than `60s`, to prevent Kubernetes from killing the pod before the `preStopHook` completes.

That being said, another solution is to use **only** `terminationGracePeriodSeconds`, and in the `preStopHook`:

- call `/healthcheck/fail` (or `/drain_listeners?graceful`, similar)
- sleep for infinity

Then, configure `terminationGracePeriodSeconds` = `<drain_max_time>`. Kubernetes will then be responsible of sending the KILL signal after `drain_max_time`, making the `preStopHook` effectively sleep for `drain_max_time`. Pro of this solution is we just have to configure one thing (`terminationGracePeriodSeconds`), con is we're mixing `terminationGracePeriodSeconds` with the drain process, so it might not be as clear (tell me what you think).

In conclusion from all of this, **my** prefered solution would be:

- define the "drain max time" in an environment variable, so users can easily change it: `DRAIN_TIME_SECONDS`
- define a bigger `terminationGracePeriodSeconds` (compared with `DRAIN_TIME_SECONDS`, e.g. `300`). If people want to have a bigger `DRAIN_TIME_SECONDS`, they must consider increasing this too
- in the `preStopHook`:
  - use `/drain_listeners?graceful`: this is just a personal preference; even if it's similar to `/healthcheck/fail`, as I said in the beginning, I feel like it's clearer to understand, because this command is **made to drain** in-flight requests
  - sleep during `DRAIN_TIME_SECONDS`. In the future, we can just wait until all in-flight requests are done, to avoid waiting for nothing. If you're happy, I can do that in a separate PR to avoid mixing too many things in this one

Tell me what you think; I'll adjust the PR accordingly :smile: sorry about the long answer, but I just wanted to explain all possibilities so we can choose the best one.

Reply:
(also see this comment on my previous MR: https://github.com/knative-extensions/net-kourier/pull/1200#issuecomment-1923168171 who might justify using `/drain_listeners` instead of `/healthcheck/fail`)

Reply:
Thank for the update! I will take a look.